### Necessary Libraries

In [475]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px
import warnings

warnings.filterwarnings("ignore")

### **Exploring The Datset**

In [476]:
df=pd.read_excel("Coffee_Shop_Sales.xlsx")

In [477]:
df.head()

transaction_id transaction_date transaction_time  transaction_qty  \
0               1       2023-01-01         07:06:11                2   
1               2       2023-01-01         07:08:56                2   
2               3       2023-01-01         07:14:04                2   
3               4       2023-01-01         07:20:24                1   
4               5       2023-01-01         07:22:41                2   

   store_id   store_location  product_id  unit_price    product_category  \
0         5  Lower Manhattan          32         3.0              Coffee   
1         5  Lower Manhattan          57         3.1                 Tea   
2         5  Lower Manhattan          59         4.5  Drinking Chocolate   
3         5  Lower Manhattan          22         2.0              Coffee   
4         5  Lower Manhattan          57         3.1                 Tea   

            product_type               product_detail  
0  Gourmet brewed coffee                  Ethiopia Rg  
1        Brewed Chai tea     Spicy Eye Opener Chai Lg  
2          Hot chocolate            Dark chocolate Lg  
3            Drip coffee  Our Old Time Diner Blend Sm  
4        Brewed Chai tea     Spicy Eye Opener Chai Lg

In [478]:
df.head()

transaction_id transaction_date transaction_time  transaction_qty  \
0               1       2023-01-01         07:06:11                2   
1               2       2023-01-01         07:08:56                2   
2               3       2023-01-01         07:14:04                2   
3               4       2023-01-01         07:20:24                1   
4               5       2023-01-01         07:22:41                2   

   store_id   store_location  product_id  unit_price    product_category  \
0         5  Lower Manhattan          32         3.0              Coffee   
1         5  Lower Manhattan          57         3.1                 Tea   
2         5  Lower Manhattan          59         4.5  Drinking Chocolate   
3         5  Lower Manhattan          22         2.0              Coffee   
4         5  Lower Manhattan          57         3.1                 Tea   

            product_type               product_detail  
0  Gourmet brewed coffee                  Ethiopia Rg  
1        Brewed Chai tea     Spicy Eye Opener Chai Lg  
2          Hot chocolate            Dark chocolate Lg  
3            Drip coffee  Our Old Time Diner Blend Sm  
4        Brewed Chai tea     Spicy Eye Opener Chai Lg

### **Data Pre-Processing**

In [479]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 149116 entries, 0 to 149115
Data columns (total 11 columns):
 #   Column            Non-Null Count   Dtype         
---  ------            --------------   -----         
 0   transaction_id    149116 non-null  int64         
 1   transaction_date  149116 non-null  datetime64[ns]
 2   transaction_time  149116 non-null  object        
 3   transaction_qty   149116 non-null  int64         
 4   store_id          149116 non-null  int64         
 5   store_location    149116 non-null  object        
 6   product_id        149116 non-null  int64         
 7   unit_price        149116 non-null  float64       
 8   product_category  149116 non-null  object        
 9   product_type      149116 non-null  object        
 10  product_detail    149116 non-null  object        
dtypes: datetime64[ns](1), float64(1), int64(4), object(5)
memory usage: 12.5+ MB


In [480]:
df.head()

transaction_id transaction_date transaction_time  transaction_qty  \
0               1       2023-01-01         07:06:11                2   
1               2       2023-01-01         07:08:56                2   
2               3       2023-01-01         07:14:04                2   
3               4       2023-01-01         07:20:24                1   
4               5       2023-01-01         07:22:41                2   

   store_id   store_location  product_id  unit_price    product_category  \
0         5  Lower Manhattan          32         3.0              Coffee   
1         5  Lower Manhattan          57         3.1                 Tea   
2         5  Lower Manhattan          59         4.5  Drinking Chocolate   
3         5  Lower Manhattan          22         2.0              Coffee   
4         5  Lower Manhattan          57         3.1                 Tea   

            product_type               product_detail  
0  Gourmet brewed coffee                  Ethiopia Rg  
1        Brewed Chai tea     Spicy Eye Opener Chai Lg  
2          Hot chocolate            Dark chocolate Lg  
3            Drip coffee  Our Old Time Diner Blend Sm  
4        Brewed Chai tea     Spicy Eye Opener Chai Lg

In [481]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 149116 entries, 0 to 149115
Data columns (total 11 columns):
 #   Column            Non-Null Count   Dtype         
---  ------            --------------   -----         
 0   transaction_id    149116 non-null  int64         
 1   transaction_date  149116 non-null  datetime64[ns]
 2   transaction_time  149116 non-null  object        
 3   transaction_qty   149116 non-null  int64         
 4   store_id          149116 non-null  int64         
 5   store_location    149116 non-null  object        
 6   product_id        149116 non-null  int64         
 7   unit_price        149116 non-null  float64       
 8   product_category  149116 non-null  object        
 9   product_type      149116 non-null  object        
 10  product_detail    149116 non-null  object        
dtypes: datetime64[ns](1), float64(1), int64(4), object(5)
memory usage: 12.5+ MB


In [482]:
num_duplicates = df.duplicated().sum()
print(f"Total number of duplicate rows in the dataset: {num_duplicates}")


Total number of duplicate rows in the dataset: 0


In [483]:
df.describe()

transaction_id               transaction_date  transaction_qty  \
count   149116.000000                         149116    149116.000000   
mean     74737.371872  2023-04-15 11:50:32.173609984         1.438276   
min          1.000000            2023-01-01 00:00:00         1.000000   
25%      37335.750000            2023-03-06 00:00:00         1.000000   
50%      74727.500000            2023-04-24 00:00:00         1.000000   
75%     112094.250000            2023-05-30 00:00:00         2.000000   
max     149456.000000            2023-06-30 00:00:00         8.000000   
std      43153.600016                            NaN         0.542509   

            store_id     product_id     unit_price  
count  149116.000000  149116.000000  149116.000000  
mean        5.342063      47.918607       3.382219  
min         3.000000       1.000000       0.800000  
25%         3.000000      33.000000       2.500000  
50%         5.000000      47.000000       3.000000  
75%         8.000000      60.000000       3.750000  
max         8.000000      87.000000      45.000000  
std         2.074241      17.930020       2.658723

**Since The Data Seems Clean Let's Deep Dive Into The Analysis**

# **Customer and Sales Analysis**

### 1. What are the peak transaction times during the day, and how do they vary by store location?

In [484]:
df.head()

transaction_id transaction_date transaction_time  transaction_qty  \
0               1       2023-01-01         07:06:11                2   
1               2       2023-01-01         07:08:56                2   
2               3       2023-01-01         07:14:04                2   
3               4       2023-01-01         07:20:24                1   
4               5       2023-01-01         07:22:41                2   

   store_id   store_location  product_id  unit_price    product_category  \
0         5  Lower Manhattan          32         3.0              Coffee   
1         5  Lower Manhattan          57         3.1                 Tea   
2         5  Lower Manhattan          59         4.5  Drinking Chocolate   
3         5  Lower Manhattan          22         2.0              Coffee   
4         5  Lower Manhattan          57         3.1                 Tea   

            product_type               product_detail  
0  Gourmet brewed coffee                  Ethiopia Rg  
1        Brewed Chai tea     Spicy Eye Opener Chai Lg  
2          Hot chocolate            Dark chocolate Lg  
3            Drip coffee  Our Old Time Diner Blend Sm  
4        Brewed Chai tea     Spicy Eye Opener Chai Lg

In [485]:
# Checking the unique values in the transaction_time column to identify format issues
unique_times = df['transaction_time'].unique()
unique_times

array([datetime.time(7, 6, 11), datetime.time(7, 8, 56),
       datetime.time(7, 14, 4), ..., datetime.time(10, 34, 30),
       datetime.time(12, 3, 24), datetime.time(16, 59, 49)], dtype=object)

In [486]:
# Combining date and time into a single datetime column
df['transaction_datetime'] = df['transaction_date'] + pd.to_timedelta(df['transaction_time'].astype(str))

# Extracting the hour from the new transaction_datetime column
df['transaction_hour'] = df['transaction_datetime'].dt.hour



In [487]:
df[['transaction_datetime','transaction_hour']]

transaction_datetime  transaction_hour
0       2023-01-01 07:06:11                 7
1       2023-01-01 07:08:56                 7
2       2023-01-01 07:14:04                 7
3       2023-01-01 07:20:24                 7
4       2023-01-01 07:22:41                 7
...                     ...               ...
149111  2023-06-30 20:18:41                20
149112  2023-06-30 20:25:10                20
149113  2023-06-30 20:31:34                20
149114  2023-06-30 20:57:19                20
149115  2023-06-30 20:57:19                20

[149116 rows x 2 columns]

In [488]:
# Aggregating transaction quantities by store location and transaction hour
peak_qty_hours = df.groupby(['store_location', 'transaction_hour'])['transaction_qty'].sum().reset_index(name='total_qty').sort_values(by='total_qty',ascending=False)

# Displaying the aggregated peak quantities
peak_qty_hours.head(10)

store_location  transaction_hour  total_qty
17   Hell's Kitchen                10       9873
15   Hell's Kitchen                 8       9653
32  Lower Manhattan                10       9452
16   Hell's Kitchen                 9       9393
31  Lower Manhattan                 9       8922
29  Lower Manhattan                 7       8850
30  Lower Manhattan                 8       8788
3           Astoria                10       7388
2           Astoria                 9       7055
1           Astoria                 8       6756

In [489]:
import plotly.express as px

# Creating the bar chart
fig_qty = px.bar(
    peak_qty_hours, 
    x='transaction_hour', 
    y='total_qty', 
    color='store_location', 
    title='<b>Total Transaction Quantities by Store Location and Hour</b>', 
    labels={'transaction_hour': 'Hour of Day', 'total_qty': 'Total Quantity Sold', 'store_location': 'Store Location'},
    text='total_qty', 
    color_discrete_sequence=px.colors.qualitative.Set1  # Apply Set1 color palette
)

# Update layout for improved readability and visibility
fig_qty.update_layout(
    title_font=dict(size=22, family='Arial', color='black'),
    font=dict(size=13, family='Arial'),
    height=470,
    width=1800,  
    plot_bgcolor='white', 
    paper_bgcolor='white',
    margin=dict(t=70, b=90, l=50, r=50),
    legend=dict(
        title='<b>Store Location</b>',  # Set clearer legend title
        title_font=dict(size=14, family='Arial', color='black'),
        font=dict(size=12, family='Arial'),
        orientation='h',
        yanchor='top',
        y=1,
        xanchor='center',
        x=0.5
    ),
    xaxis=dict(
        title='<b>Hour of Day</b>',
        title_font=dict(size=14, family='Arial', weight='bold'),
        tickangle=-45,
        tickmode='array',
        tickvals=list(range(24)),
        ticktext=[f'{i}:00' for i in range(24)],
        tickfont=dict(size=12, family='Arial', weight='bold'),
        showgrid=True,
        gridcolor='lightgray',
        zeroline=True,
        zerolinewidth=1,
        zerolinecolor='gray'
    ),
    yaxis=dict(
        title='<b>Total Quantity Sold</b>',
        title_font=dict(size=14, family='Arial', weight='bold'),
        tickfont=dict(size=12, family='Arial', weight='bold'),
        showgrid=True,
        gridcolor='lightgray',
        zeroline=True,
        zerolinewidth=1,
        zerolinecolor='gray'
    ),
    barmode='group',
    showlegend=True
)

# Adjust text visibility and bar appearance (add thin white border)
fig_qty.update_traces(
    texttemplate='%{text:.3s}',  # Format text on bars (shortened if necessary)
    textposition='outside',
    marker_line_width=1,  # Thin white border around bars
    marker_line_color='white',  # Set border color to white
    textfont=dict(size=12, color='black', family='Arial', weight='bold')
)

# Improve hover template for better clarity
fig_qty.update_traces(
    #hovertemplate='<b>Store Location:</b> %{color}<br><b>Hour:</b> %{x}:00<br><b>Quantity Sold:</b> %{y}<extra></extra>'
)

# Show the plot
fig_qty.show()


In [490]:
fig_qty.write_html("Total Transaction Quantities by Store Location and Hour.html")

---

### **Interpretation of the Data**
The bar chart reveals distinct peak transaction times for three store locations: Hell's Kitchen, Lower Manhattan, and Astoria. Hell's Kitchen experiences its highest transaction volume at 10:00 AM with 9.87k transactions, followed by 9:00 AM and 8:00 AM. Lower Manhattan peaks at 10:00 AM with 8.85k transactions, with other high transaction times at 9:00 AM and 8:00 AM. Astoria, on the other hand, also sees its peak at 10:00 AM with 7.39k transactions, with significant activity also at 8:00 AM and 9:00 AM. Additionally, Hell's Kitchen and Lower Manhattan show a steady decline in transactions after the morning peak, with a slight increase around lunchtime. Astoria, however, maintains a relatively consistent transaction volume throughout the day, with another noticeable peak in the late afternoon.

---
### **Business Insights and Suggestions**
For shop owners, these insights suggest several opportunities for growth and improvement. Hell's Kitchen and Lower Manhattan stores should consider increasing staffing and inventory during the early morning hours to accommodate the high transaction volumes. Marketing efforts could be targeted towards morning commuters to capitalize on this peak period. For the Astoria location, the focus should be on the evening hours, potentially offering promotions or events to attract more customers during this time. Additionally, analyzing the factors contributing to these peak times could help replicate successful strategies across other locations. For all locations, maintaining a steady inventory and staff presence during secondary peak times, such as lunchtime and late afternoon, can help capture additional sales and improve customer satisfaction.

---

### 2.Which store location has the highest total sales revenue ?

In [491]:
df.head()

transaction_id transaction_date transaction_time  transaction_qty  \
0               1       2023-01-01         07:06:11                2   
1               2       2023-01-01         07:08:56                2   
2               3       2023-01-01         07:14:04                2   
3               4       2023-01-01         07:20:24                1   
4               5       2023-01-01         07:22:41                2   

   store_id   store_location  product_id  unit_price    product_category  \
0         5  Lower Manhattan          32         3.0              Coffee   
1         5  Lower Manhattan          57         3.1                 Tea   
2         5  Lower Manhattan          59         4.5  Drinking Chocolate   
3         5  Lower Manhattan          22         2.0              Coffee   
4         5  Lower Manhattan          57         3.1                 Tea   

            product_type               product_detail transaction_datetime  \
0  Gourmet brewed coffee                  Ethiopia Rg  2023-01-01 07:06:11   
1        Brewed Chai tea     Spicy Eye Opener Chai Lg  2023-01-01 07:08:56   
2          Hot chocolate            Dark chocolate Lg  2023-01-01 07:14:04   
3            Drip coffee  Our Old Time Diner Blend Sm  2023-01-01 07:20:24   
4        Brewed Chai tea     Spicy Eye Opener Chai Lg  2023-01-01 07:22:41   

   transaction_hour  
0                 7  
1                 7  
2                 7  
3                 7  
4                 7

In [492]:
# Calculating total sales revenue by store location
# Adding a new column for total revenue
df['total_revenue'] = df['transaction_qty'] * df['unit_price']
# Grouping by store location to get total revenue
total_revenue_by_store = df.groupby('store_location')['total_revenue'].sum().reset_index().sort_values(by="total_revenue",ascending=False)
total_revenue_by_store

store_location  total_revenue
1   Hell's Kitchen      236511.17
0          Astoria      232243.91
2  Lower Manhattan      230057.25

In [493]:
# Creating a bar chart to visualize total sales revenue by store location
fig_revenue = px.bar(total_revenue_by_store, 
                     x='store_location', 
                     y='total_revenue', 
                     title='Total Sales Revenue by Store Location', 
                     labels={'store_location': 'Store Location', 'total_revenue': 'Total Revenue'},
                     text='total_revenue', 
                     color='store_location',  
                     color_discrete_sequence=px.colors.qualitative.Set1)  # Set1 color palette

# Find the store with the highest revenue
max_revenue_row = total_revenue_by_store.loc[total_revenue_by_store['total_revenue'].idxmax()]
max_revenue = max_revenue_row['total_revenue']
max_revenue_location = max_revenue_row['store_location']

# Add annotation with an arrow pointing to the highest bar
fig_revenue.add_annotation(
    x=max_revenue_location, 
    y=max_revenue, 
    text=f"Highest: {max_revenue:.2f}",
    showarrow=True,
    arrowhead=2,  
    arrowsize=1, 
    ax=-70,  
    ay=-50,  
    font=dict(size=12, color='black', family='Arial', weight='bold'),
    bgcolor='white',
    borderpad=4,
    bordercolor='black',
    standoff=16
    
)

# Updating layout for better visibility and professional finish
fig_revenue.update_layout(
    title_text='<b>Total Sales Revenue by Store Location</b>',
    title_font=dict(size=18, family='Arial', color='black'),
    font=dict(size=11, family='Arial'),
    height=470,
    plot_bgcolor='white',
    margin=dict(t=50, b=50, l=50, r=50),
    
    yaxis=dict(
        title='Total Revenue',
        showgrid=True,
        gridcolor='lightgray',  # Light gridlines for readability
    )
)

# Adjust text visibility and bar appearance
fig_revenue.update_traces(
    texttemplate='%{text:.3s}',  # Format text on bars
    textposition='outside',      # Position text outside for better visibility
    marker_line_width=1.5,       # Add subtle borders for bars
    textfont=dict(size=12, color='black', family='Arial', weight='bold'),  # Make text bold and more visible
)

# Show the plot
fig_revenue.show()


In [494]:
fig_revenue.write_html("fig_revenue.html")

---

### **Interpretation of the Data**
The bar chart shows that Hell's Kitchen has the highest total sales revenue, amounting to $236,611.17. Astoria follows closely with $232,000, and Lower Manhattan has a total revenue of $230,000. 

---
### **Business Insights and Suggestions**
These insights suggest that Hell's Kitchen is currently the top-performing store location in terms of sales. Shop owners could focus on maintaining and enhancing the customer experience at Hell's Kitchen to sustain this lead. Additionally, analyzing the factors contributing to Hell's Kitchen's success could provide valuable strategies to boost sales at the other locations. For Astoria and Lower Manhattan, targeted marketing campaigns and promotions could help close the gap and potentially surpass Hell's Kitchen in total sales.

---

### 3. What are the best-selling product categories and types across different store locations?

In [495]:
df.head()

transaction_id transaction_date transaction_time  transaction_qty  \
0               1       2023-01-01         07:06:11                2   
1               2       2023-01-01         07:08:56                2   
2               3       2023-01-01         07:14:04                2   
3               4       2023-01-01         07:20:24                1   
4               5       2023-01-01         07:22:41                2   

   store_id   store_location  product_id  unit_price    product_category  \
0         5  Lower Manhattan          32         3.0              Coffee   
1         5  Lower Manhattan          57         3.1                 Tea   
2         5  Lower Manhattan          59         4.5  Drinking Chocolate   
3         5  Lower Manhattan          22         2.0              Coffee   
4         5  Lower Manhattan          57         3.1                 Tea   

            product_type               product_detail transaction_datetime  \
0  Gourmet brewed coffee                  Ethiopia Rg  2023-01-01 07:06:11   
1        Brewed Chai tea     Spicy Eye Opener Chai Lg  2023-01-01 07:08:56   
2          Hot chocolate            Dark chocolate Lg  2023-01-01 07:14:04   
3            Drip coffee  Our Old Time Diner Blend Sm  2023-01-01 07:20:24   
4        Brewed Chai tea     Spicy Eye Opener Chai Lg  2023-01-01 07:22:41   

   transaction_hour  total_revenue  
0                 7            6.0  
1                 7            6.2  
2                 7            9.0  
3                 7            2.0  
4                 7            6.2

In [496]:
## Top Selling Product Categories based on Location
category_summary = df.groupby(['store_location', 'product_category'])['transaction_qty'].sum().sort_values(ascending=False).reset_index()
category_summary 

store_location    product_category  transaction_qty
0    Hell's Kitchen              Coffee            30220
1           Astoria              Coffee            29655
2   Lower Manhattan              Coffee            29375
3           Astoria                 Tea            24048
4    Hell's Kitchen                 Tea            22980
5   Lower Manhattan                 Tea            22709
6   Lower Manhattan              Bakery             8040
7    Hell's Kitchen              Bakery             7678
8           Astoria              Bakery             7496
9           Astoria  Drinking Chocolate             6351
10   Hell's Kitchen  Drinking Chocolate             5672
11  Lower Manhattan  Drinking Chocolate             5434
12  Lower Manhattan            Flavours             4709
13   Hell's Kitchen            Flavours             3596
14          Astoria            Flavours             2206
15   Hell's Kitchen        Coffee beans              790
16  Lower Manhattan        Coffee beans              536
17          Astoria        Coffee beans              502
18   Hell's Kitchen           Loose Tea              485
19  Lower Manhattan           Loose Tea              381
20  Lower Manhattan             Branded              378
21          Astoria           Loose Tea              344
22          Astoria             Branded              279
23   Hell's Kitchen  Packaged Chocolate              197
24  Lower Manhattan  Packaged Chocolate              180
25   Hell's Kitchen             Branded              119
26          Astoria  Packaged Chocolate              110

In [497]:
product_type_summary = df.groupby(['store_location', 'product_type'])['transaction_qty'].sum().reset_index()
product_type_summary = product_type_summary.sort_values(by='transaction_qty', ascending=False).groupby('store_location').head(10)
product_type_summary

store_location           product_type  transaction_qty
4           Astoria        Brewed Chai tea             9306
29   Hell's Kitchen       Barista Espresso             9064
13          Astoria  Gourmet brewed coffee             8938
33   Hell's Kitchen        Brewed Chai tea             8755
71  Lower Manhattan  Gourmet brewed coffee             8563
58  Lower Manhattan       Barista Espresso             8534
42   Hell's Kitchen  Gourmet brewed coffee             8472
62  Lower Manhattan        Brewed Chai tea             8189
0           Astoria       Barista Espresso             7345
17          Astoria          Hot chocolate             6351
6           Astoria      Brewed herbal tea             5953
61  Lower Manhattan       Brewed Black tea             5907
3           Astoria       Brewed Black tea             5891
64  Lower Manhattan      Brewed herbal tea             5820
46   Hell's Kitchen          Hot chocolate             5672
32   Hell's Kitchen       Brewed Black tea             5664
35   Hell's Kitchen      Brewed herbal tea             5555
75  Lower Manhattan          Hot chocolate             5434
10          Astoria            Drip coffee             4577
51   Hell's Kitchen  Organic brewed coffee             4518
25          Astoria  Premium brewed coffee             4451
22          Astoria  Organic brewed coffee             4344
39   Hell's Kitchen            Drip coffee             4342
83  Lower Manhattan  Premium brewed coffee             4156
80  Lower Manhattan  Organic brewed coffee             4150
68  Lower Manhattan            Drip coffee             3972
54   Hell's Kitchen  Premium brewed coffee             3824
85  Lower Manhattan                  Scone             3656
56   Hell's Kitchen                  Scone             3520
27          Astoria                  Scone             3289

In [498]:
import plotly.graph_objects as go
from plotly.subplots import make_subplots
import plotly.express as px

# Define color palette for consistency
color_palette = px.colors.qualitative.Set1

# Create subplot layout with 2 rows and 1 column
fig_combined = make_subplots(
    rows=2, cols=1,
    shared_xaxes=False,
    row_heights=[0.45, 0.45],  # Adjust row heights
    subplot_titles=(
        '<b>Best-Selling Product Categories by Store Location (Units Sold)</b>',
        '<b>Top 10 Product Types by Store Location (Units Sold)</b>'
    ),
    vertical_spacing=0.3  # Adjust spacing between the plots
)

# Create a consistent color map for store locations to ensure the same color is used across both plots
location_colors = {location: color_palette[i % len(color_palette)] for i, location in enumerate(category_summary['store_location'].unique())}

# Plot 1: Top-Selling Product Categories by Store Location
for location in category_summary['store_location'].unique():
    filtered_data = category_summary[category_summary['store_location'] == location]
    fig_combined.add_trace(
        go.Bar(
            x=filtered_data['product_category'],
            y=filtered_data['transaction_qty'],
            name=location,
            text=filtered_data['transaction_qty'],
            texttemplate='%{text:.3s}',  # Format text on bars
            textposition='outside',
            marker=dict(color=location_colors[location]),
            legendgroup=location,  # Group all bars for a location together
            hovertemplate="<b>%{x}</b><br>Qty Sold: %{y}<br>Location: %{customdata[0]}<br><extra></extra>",
            customdata=filtered_data[['store_location']].values,  # Pass location to hover text
            showlegend=True  # Ensure legend is shown only for the first plot
        ),
        row=1, col=1
    )

# Plot 2: Top 10 Product Types by Store Location
for location in product_type_summary['store_location'].unique():
    filtered_data = product_type_summary[product_type_summary['store_location'] == location]
    fig_combined.add_trace(
        go.Bar(
            x=filtered_data['product_type'],
            y=filtered_data['transaction_qty'], 
            name=location,
            text=filtered_data['transaction_qty'],
            texttemplate='%{text:.3s}',  # Maintain the same text template for the second plot
            textposition='outside',
            marker=dict(color=location_colors[location]),
            legendgroup=location,  # Group all bars for a location together
            hovertemplate="<b>%{x}</b><br>Qty Sold: %{y}<br>Location: %{customdata[0]}<br><extra></extra>",
            customdata=filtered_data[['store_location']].values,  # Pass location to hover text
            showlegend=False  # Hide legend for the second plot
        ),
        row=2, col=1
    )

# Update layout for the combined figure
fig_combined.update_layout(
    title_text='<b>Sales Analysis by Product Category and Product Type</b>',
    title_font=dict(size=22, family='Arial', color='black'),
    font=dict(size=14, family='Arial'),
    height=1000,  
    plot_bgcolor='white',  
    margin=dict(t=60, b=50, l=50, r=50),
    legend=dict(
        title='<b>Store Location</b>',  # Set clearer legend title
        title_font=dict(size=14, family='Arial', color='black'),
        font=dict(size=12, family='Arial'),
        orientation='h',
        yanchor='top',
        y=.98,
        xanchor='center',
        x=0.5
    )
)

# Update x- and y-axis titles for both subplots
fig_combined.update_xaxes(
    title_text="Product Category",
    row=1, col=1,
    tickangle=-45,
    showgrid=True,  
    gridcolor='lightgray'  
)

fig_combined.update_yaxes(
    title_text="Quantity Sold",
    row=1, col=1,
    showgrid=True,  
    gridcolor='lightgray'  
)

fig_combined.update_xaxes(
    title_text="Quantity Sold",
    row=2, col=1,
    showgrid=True,  
    gridcolor='lightgray'  
)

fig_combined.update_yaxes(
    title_text="Product Type",
    row=2, col=1,
    categoryorder='total ascending',
    showgrid=True,  
    gridcolor='lightgray'  
)

# Show the combined plot
fig_combined.show()


In [499]:
fig_combined.write_html("sales_analysis_by_product_cat_type.html")

---

### **Interpretation of the Data**  
The bar charts reveal key product sales across three locations. **Coffee** is the top-selling category in all locations, with Hell's Kitchen leading at 30.2k units, followed by Astoria (29.7k) and Lower Manhattan (29.4k). **Tea** is also popular, with Astoria leading at 24.0k units. Other notable categories include **Drinking Chocolate**, **Flavours**, **Coffee Beans**, and **Packaged Chocolate**. The top-selling product types are **Brewed Chai Tea** (Astoria: 9.31k, Hell's Kitchen: 8.76k, Lower Manhattan: 8.19k), followed by **Gourmet Brewed Coffee** and **Barista Espresso**.

---

### **Business Insights & Suggestions**  
Shop owners should prioritize **Coffee** and **Tea**, ensuring these categories are well-stocked and promoted. Hell's Kitchen and Astoria, in particular, can continue focusing on these categories, with opportunities to introduce new flavors or seasonal variations to attract customers. Highlighting top-selling product types like **Brewed Chai Tea** and **Gourmet Brewed Coffee** can drive sales and enhance customer loyalty. Promotions and loyalty programs around these items could increase retention. Additionally, exploring the appeal of other categories like **Drinking Chocolate**, **Flavours**, and **Branded products** can help attract a broader customer base and provide a balanced product offering.

---

### 4. How do these best-sellers align with the categories that contribute most to overall profit?

In [500]:
profit_summary = df.groupby('product_category')['total_revenue'].sum().reset_index().sort_values(by='total_revenue',ascending=False)

In [501]:
# Group by store location and product category, then sum the total revenue
profit_summary = (
    df.groupby(['store_location', 'product_category'])['total_revenue']
    .sum()
    .reset_index()
    .sort_values(by='total_revenue', ascending=False)
)

# Display the result
profit_summary.head()


store_location product_category  total_revenue
11   Hell's Kitchen           Coffee       91222.65
2           Astoria           Coffee       89744.30
20  Lower Manhattan           Coffee       88985.50
8           Astoria              Tea       67839.90
17   Hell's Kitchen              Tea       64701.30

In [502]:
# Creating the side-by-side bar chart for total profit contribution by product category for each store
fig = px.bar(
    profit_summary,
    x='product_category',
    y='total_revenue',
    color='store_location',  # Group bars by store location
    barmode='group',  # Side-by-side bars
    title='<b>Total Profit Contribution by Product Category and Store Location</b>',
    labels={'total_revenue': 'Total Profit', 'product_category': 'Product Category'},
    color_discrete_sequence=px.colors.qualitative.Set1,
    text='total_revenue'  # Display total revenue on top of the bars
)

# Update layout to improve readability and presentation
fig.update_layout(
    title_font=dict(size=20, family='Arial', color='black'),
    font=dict(size=13, family='Arial'),
    height=500,
    width=1000,
    plot_bgcolor='white',
    margin=dict(t=60, b=100, l=50, r=50), 
    showlegend=True,
    legend=dict(
        title='Store Location',
        orientation='h',
        x=0.5,
        xanchor='center',
        y=0.95,
    ),
    xaxis=dict(
        title='Product Category',
        tickangle=-45,  # Rotate x-axis labels for readability
        tickfont=dict(size=12, family='Arial', weight='bold'),
        showgrid=True,
        gridcolor='lightgray',
    ),
    yaxis=dict(
        title='Total Profit',
        showgrid=True,
        gridcolor='lightgray',
    )
)

# Enhance text visibility and bar appearance
fig.update_traces(
    texttemplate='%{text:.4s}', 
    textposition='outside',
    marker_line_width=1.5,
    textfont=dict(size=13, color='black', family='Arial', weight='bold')
)

# Show the plot
fig.show()


---

### **Interpretation of the Data**  
Coffee and Tea are the highest profit drivers, consistent with their top sales. Hell's Kitchen leads in coffee profit ($91.22k), followed by Astoria and Lower Manhattan. Coffee Beans,Bakery contributes moderately, while Flavours, and Packaged Chocolate show minimal profit.

---

### **Business Insights & Suggestions**  

Focus on maintaining strong stock and promotions for Coffee and Tea, as they are top profit drivers. Enhance Bakery sales through strategic bundling or seasonal promotions. Promote lower-profit items like Flavours and Packaged Chocolate through loyalty programs or add-ons. Tailor marketing strategies by location—e.g., exclusive coffee products for Hell's Kitchen and tea-focused campaigns for Astoria—to maximize profitability across all stores.

---


### 5.Which product details or specific items generate the highest revenue and which have the highest transaction frequency?

In [503]:
df.head(2)

transaction_id transaction_date transaction_time  transaction_qty  \
0               1       2023-01-01         07:06:11                2   
1               2       2023-01-01         07:08:56                2   

   store_id   store_location  product_id  unit_price product_category  \
0         5  Lower Manhattan          32         3.0           Coffee   
1         5  Lower Manhattan          57         3.1              Tea   

            product_type            product_detail transaction_datetime  \
0  Gourmet brewed coffee               Ethiopia Rg  2023-01-01 07:06:11   
1        Brewed Chai tea  Spicy Eye Opener Chai Lg  2023-01-01 07:08:56   

   transaction_hour  total_revenue  
0                 7            6.0  
1                 7            6.2

In [504]:
top_10_product_detail_by_revenue=df.groupby("product_detail")["total_revenue"].sum().sort_values(ascending=False).reset_index().head(10)

In [505]:
top_10_product_detail_by_transaction_qty=df.groupby("product_detail")["transaction_qty"].count().sort_values(ascending=False).reset_index().head(10)

In [506]:
# Sorting both dataframes for a clear ranking
top_10_product_detail_by_revenue = top_10_product_detail_by_revenue.sort_values(by="total_revenue", ascending=True)
top_10_product_detail_by_transaction_qty = top_10_product_detail_by_transaction_qty.sort_values(by="transaction_qty", ascending=True)

color_palette = px.colors.qualitative.Set1

# Creating the subplot structure with two rows
fig_combined = make_subplots(
    rows=2, cols=1,
    shared_xaxes=False,
    row_heights=[0.5, 0.5],
    subplot_titles=(
        "<b>Top 10 Product Details by Total Revenue</b>",
        "<b>Top 10 Product Details by Transaction Quantity</b>"
    ),
    vertical_spacing=0.2
)

# First chart: Top 10 product details by total revenue
fig_combined.add_trace(
    go.Bar(
        x=top_10_product_detail_by_revenue['total_revenue'],
        y=top_10_product_detail_by_revenue['product_detail'],
        text=top_10_product_detail_by_revenue['total_revenue'],
        orientation='h',
        marker_color=color_palette[0],
        texttemplate='%{text:.3s}',
        textposition='outside'
    ),
    row=1, col=1
)

# Second chart: Top 10 product details by transaction quantity
fig_combined.add_trace(
    go.Bar(
        x=top_10_product_detail_by_transaction_qty['transaction_qty'],
        y=top_10_product_detail_by_transaction_qty['product_detail'],
        text=top_10_product_detail_by_transaction_qty['transaction_qty'],
        orientation='h',
        marker_color=color_palette[1],
        texttemplate='%{text:.3s}',
        textposition='outside'
    ),
    row=2, col=1
)

# Update layout for a presentation-ready appearance
fig_combined.update_layout(
    height=800,
    width=900,
    title_text="<b>Top 10 Products by Revenue and Transaction Volume</b>",
    title_font=dict(size=22, family='Arial', color='black'),
    font=dict(size=12, family='Arial'),
    plot_bgcolor='white',
    margin=dict(t=80, b=50, l=50, r=50),
    showlegend=False,
)

# Customize x-axis and y-axis titles and style
fig_combined.update_xaxes(
    title_text="Total Revenue",
    row=1, col=1,
    showgrid=True,
    gridcolor='lightgray'
)
fig_combined.update_yaxes(
    title_text="Product Detail",
    row=1, col=1,
    showgrid=True,
    gridcolor='lightgray'
)
fig_combined.update_xaxes(
    title_text="Transaction Quantity",
    row=2, col=1,
    showgrid=True,
    gridcolor='lightgray'
)
fig_combined.update_yaxes(
    title_text="Product Detail",
    row=2, col=1,
    showgrid=True,
    gridcolor='lightgray'
)

# Show the combined plot
fig_combined.show()


In [507]:
fig_combined.write_html("top_10_product_details.html")

---

### **Interpretation of the Data**  
The charts reveal key trends in both revenue generation and transaction frequency. The top chart shows that **Sustainably Grown Organic Lg** leads in revenue, generating $21.2k, closely followed by **Dark Chocolate Lg** at $21.0k, and **Latte Rg** at $19.1k. Other notable high-revenue products include **Cappuccino Lg**, **Morning Sunrise Chai Lg**, and **Jamaican Coffee River Lg**. The second chart, which highlights transaction volume, shows **Chocolate Croissant** as the most frequently purchased item, with 3.08k transactions, followed by **Earl Grey Rg** at 3.05k and **Dark Chocolate Lg** at 3.03k. Other popular products in terms of transaction frequency include **Morning Sunrise Chai Rg**, **Colombian Medium Roast Rg**, and **Latte**.

---

### **Business Insights & Suggestions**  
Shop owners should focus on the high-revenue products like **Sustainably Grown Organic Lg** and **Dark Chocolate Lg**, ensuring they are well-stocked and prominently displayed to maintain sales. Additionally, promoting frequently purchased items such as **Chocolate Croissant** and **Earl Grey Rg** can drive repeat business and increase transaction volume. Introducing bundle deals or loyalty programs around these popular products could also help enhance customer retention. Offering new or seasonal variations of these top products might attract both new customers and keep existing ones engaged.

---

### 6.How does the average unit price impact the total revenue and quantity sold for different product categories?

In [508]:
df.head()

transaction_id transaction_date transaction_time  transaction_qty  \
0               1       2023-01-01         07:06:11                2   
1               2       2023-01-01         07:08:56                2   
2               3       2023-01-01         07:14:04                2   
3               4       2023-01-01         07:20:24                1   
4               5       2023-01-01         07:22:41                2   

   store_id   store_location  product_id  unit_price    product_category  \
0         5  Lower Manhattan          32         3.0              Coffee   
1         5  Lower Manhattan          57         3.1                 Tea   
2         5  Lower Manhattan          59         4.5  Drinking Chocolate   
3         5  Lower Manhattan          22         2.0              Coffee   
4         5  Lower Manhattan          57         3.1                 Tea   

            product_type               product_detail transaction_datetime  \
0  Gourmet brewed coffee                  Ethiopia Rg  2023-01-01 07:06:11   
1        Brewed Chai tea     Spicy Eye Opener Chai Lg  2023-01-01 07:08:56   
2          Hot chocolate            Dark chocolate Lg  2023-01-01 07:14:04   
3            Drip coffee  Our Old Time Diner Blend Sm  2023-01-01 07:20:24   
4        Brewed Chai tea     Spicy Eye Opener Chai Lg  2023-01-01 07:22:41   

   transaction_hour  total_revenue  
0                 7            6.0  
1                 7            6.2  
2                 7            9.0  
3                 7            2.0  
4                 7            6.2

In [509]:
# Grouping the data by product category
summary_data = df.groupby('product_category').agg(
    average_unit_price=('unit_price', 'mean'),
    total_Revenue=('total_revenue', 'sum'),
    total_quantity_sold=('transaction_qty', 'sum')
).reset_index()
summary_data

product_category  average_unit_price  total_Revenue  total_quantity_sold
0              Bakery            3.551682       82315.64                23214
1             Branded           17.720214       13607.00                  776
2              Coffee            3.023646      269952.45                89250
3        Coffee beans           21.018397       40085.25                 1828
4  Drinking Chocolate            4.148827       72416.00                17457
5            Flavours            0.800000        8408.80                10511
6           Loose Tea            9.267438       11213.60                 1210
7  Packaged Chocolate            9.050595        4407.64                  487
8                 Tea            2.817121      196405.95                69737

In [510]:
fig = px.scatter(
    summary_data,
    x='average_unit_price',
    y='total_Revenue',
    size='total_quantity_sold',
    color='product_category',
    hover_name='product_category',
    title='<b>Impact of Average Unit Price on Total Revenue and Quantity Sold by Product Category</b>',
    labels={
        'average_unit_price': 'Average Unit Price', 
        'total_sales': 'Total Sales'
    },
    color_discrete_sequence=px.colors.qualitative.Set1  
)

fig.update_layout(
    title_font=dict(size=20, family='Arial', color='black'),
    font=dict(size=12, family='Arial'),
    height=600,
    width=1200,
    plot_bgcolor='white',
    margin=dict(t=60, b=50, l=50, r=50),
    xaxis=dict(
        title='Average Unit Price',
        showgrid=True,
        gridcolor='lightgray'
    ),
    yaxis=dict(
        title='Total Revenue',
        showgrid=True,
        gridcolor='lightgray'
    ),
    showlegend=True  # Keep the legend visible
)

# Adjust marker sizes to ensure visibility without overwhelming the chart
fig.update_traces(
    marker=dict(
        sizemode='area',
        sizeref=2.5*max(summary_data['total_quantity_sold'])/(60**2),  # Optimized for better balance
        sizemin=9,  # Ensure small points are visible without being too large
        line=dict(width=1.5, color='black')
    )
)

# Show the plot
fig.show()


In [511]:
fig.write_html("Impact of Average Unit Price on Total Sales and Quantity Sold by Product Category.html")

---

### **Interpretation of Data:**
The scatter plot shows that average unit price significantly impacts both total revenue and quantity sold across product categories. Lower-priced categories like Tea and Coffee generate the highest total revenues due to their large quantities sold, with Tea around $200k and Coffee around $300k in revenue. In contrast, Coffee Beans, with the highest average unit price, generates relatively low total revenue and quantity sold, indicating lower demand at this price point. Other affordable categories like Bakery and Drinking Chocolate show moderate revenue levels, suggesting that lower prices encourage higher sales volumes.

---

### **Business Insights and Suggestions:**
The data suggests that lower-priced products drive higher sales volumes, resulting in significant total revenue. Business owners should focus on competitively priced, high-demand categories like Coffee and Tea to maximize revenue. For higher-priced categories like Coffee Beans, which show limited sales, strategies such as bundling, discounts, or smaller, more affordable package sizes could help boost sales. The steady performance of Bakery and Drinking Chocolate presents an opportunity to introduce new flavors or promotions to attract more customers. A balanced mix of premium and value-oriented products can help expand the customer base and increase overall profitability.

---

### 7.What trends can be observed in transaction volume over the months, and are there any noticeable patterns related to specific days of the week?

In [512]:
## Extracting Month and Day
#df['month_name'] = df['transaction_date'].dt.month_name()
df['month'] = df['transaction_date'].dt.to_period('M').astype(str)
df['day_of_week'] = df['transaction_date'].dt.day_name()



In [513]:
monthly_volume = df.groupby('month')['transaction_qty'].sum().reset_index()
monthly_volume 

month  transaction_qty
0  2023-01            24870
1  2023-02            23550
2  2023-03            30406
3  2023-04            36469
4  2023-05            48233
5  2023-06            50942

In [514]:
#Grouping by Week name to calculate total transaction volume
weekly_volume = df.groupby('day_of_week')['transaction_qty'].sum().reset_index()
weekly_volume

day_of_week  transaction_qty
0      Friday            31207
1      Monday            31231
2    Saturday            29614
3      Sunday            30182
4    Thursday            31162
5     Tuesday            30449
6   Wednesday            30625

In [515]:
days_order = ['Saturday', 'Sunday','Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday']
weekly_volume['day_of_week'] = pd.Categorical(weekly_volume['day_of_week'], categories=days_order, ordered=True)
weekly_volume = weekly_volume.sort_values('day_of_week')


In [516]:
from plotly.subplots import make_subplots
import plotly.graph_objects as go
import plotly.express as px

# Creating the figure with two subplots, one for each bar chart
fig_combined = make_subplots(
    rows=2, cols=1,
    shared_xaxes=False,
    row_heights=[0.5, 0.5],
    subplot_titles=(
        "<b>Total Transaction Volume by Month</b>",
        "<b>Total Transaction Volume by Day of the Week</b>"
    ),
    vertical_spacing=0.2
)

# Colors for the line chart
line_color = 'black'

# Bar chart for total transaction volume by month
fig_combined.add_trace(
    go.Bar(
        x=monthly_volume['month'],
        y=monthly_volume['transaction_qty'],
        text=monthly_volume['transaction_qty'],
        name="Monthly Volume",
        marker_color=px.colors.qualitative.Set1[0],
        texttemplate='%{text:.3s}',
        textposition='outside'
    ),
    row=1, col=1
)

# Line chart for total transaction volume by month
fig_combined.add_trace(
    go.Scatter(
        x=monthly_volume['month'],
        y=monthly_volume['transaction_qty'],
        name="Monthly Trend",
        mode="lines+markers",
        line=dict(color=line_color, width=2),
        marker=dict(size=6)
    ),
    row=1, col=1
)

# Bar chart for total transaction volume by day of the week
fig_combined.add_trace(
    go.Bar(
        x=weekly_volume['day_of_week'],
        y=weekly_volume['transaction_qty'],
        text=weekly_volume['transaction_qty'],
        name="Weekly Volume",
        marker_color=px.colors.qualitative.Set1[1],
        texttemplate='%{text:.3s}',
        textposition='outside'
    ),
    row=2, col=1
)

# Line chart for total transaction volume by day of the week
fig_combined.add_trace(
    go.Scatter(
        x=weekly_volume['day_of_week'],
        y=weekly_volume['transaction_qty'],
        name="Weekly Trend",
        mode="lines+markers",
        line=dict(color=line_color, width=2),
        marker=dict(size=6)
    ),
    row=2, col=1
)

# Update layout for a polished appearance
fig_combined.update_layout(
    height=800,
    width=900,
    title_text="<b>Combined Transaction Volume Analysis</b>",
    title_font=dict(size=22, family='Arial', color='black'),
    font=dict(size=12, family='Arial'),
    plot_bgcolor='white',
    margin=dict(t=80, b=50, l=50, r=50),
    showlegend=False,
)

# Update x-axis and y-axis titles and style
fig_combined.update_xaxes(
    title_text="Month",
    row=1, col=1,
    tickangle=-45,
    showgrid=True,
    gridcolor='lightgray'
)
fig_combined.update_yaxes(
    title_text="Total Transaction Volume",
    row=1, col=1,
    showgrid=True,
    gridcolor='lightgray'
)
fig_combined.update_xaxes(
    title_text="Day of the Week",
    row=2, col=1,
    tickangle=-45,
    showgrid=True,
    gridcolor='lightgray'
)
fig_combined.update_yaxes(
    title_text="Total Transaction Volume",
    row=2, col=1,
    showgrid=True,
    gridcolor='lightgray'
)

# Show the combined plot
fig_combined.show()


In [517]:
fig_combined.write_html("Month_week_analysis.html")


---

### **Interpretation of the Data**  
The charts reveal key trends in transaction volume. The monthly chart shows a steady increase from January 2023 (24.9k transactions) to June 2023 (50.9k transactions), suggesting a growing customer base or rising purchasing activity. The weekly chart indicates stable transaction volumes, with slight peaks on Monday, Thursday, and Friday (~31.2k transactions) and the lowest volume on Saturday (29.6k transactions).

---

### **Business Insights & Suggestions**  
To meet the rising demand indicated by the monthly growth, shop owners should consider increasing inventory and staffing. Marketing campaigns aligned with this trend can further enhance impact. Monday, Thursday, and Friday are ideal for new promotions or special offers, leveraging higher customer activity. Addressing the lower Saturday volume with weekend specials or targeted ads could help boost sales on this day.

---



### 8. What product categories experience the highest growth rate in sales over different months?

In [518]:
#Aggregate monthly sales quantities by product category
monthly_sales = df.groupby(['month', 'product_category'])['total_revenue'].sum().reset_index()

#Sort data for calculating month-over-month growth rate
monthly_sales['month'] = pd.to_datetime(monthly_sales['month'])
monthly_sales.sort_values(by=['product_category', 'month'], inplace=True)

# Calculate month-over-month growth rate for each category
monthly_sales['sales_growth_rate'] = monthly_sales.groupby('product_category')['total_revenue'].pct_change()

# Filter out rows where growth rate is NaN 
growth_data = monthly_sales.dropna(subset=['sales_growth_rate'])

# Display the data with growth rates
growth_data_sorted = growth_data.sort_values(by='sales_growth_rate', ascending=False)
growth_data_sorted.head()


month    product_category  total_revenue  sales_growth_rate
34 2023-04-01  Packaged Chocolate         873.23           0.484580
19 2023-03-01             Branded        1801.00           0.458300
41 2023-05-01            Flavours        1905.60           0.343486
44 2023-05-01                 Tea       44539.85           0.335249
40 2023-05-01  Drinking Chocolate       16319.75           0.330405

In [519]:
# Creating a copy of growth_data with sales_growth_rate multiplied by 100 for hover display
growth_data['sales_growth_rate_display'] = growth_data['sales_growth_rate'] * 100

# Creating the line chart to visualize month-over-month sales growth rate by product category
fig = px.line(
    growth_data, 
    x='month',
    y='sales_growth_rate_display',  
    color='product_category',
    title='<b>Month-over-Month Sales Growth Rate by Product Category</b>',
    labels={
        'month': 'Month',
        'sales_growth_rate_display': 'Growth Rate (%)',
        'product_category': 'Product Category'
    },
    color_discrete_sequence=px.colors.qualitative.Set1  
)

# Update layout to fix the x-axis month repetition and keep line subtle
fig.update_layout(
    title_font=dict(size=20, family='Arial', color='black'),
    font=dict(size=12, family='Arial'),
    height=350,
    width=1200,
    plot_bgcolor='white',
    margin=dict(t=60, b=50, l=50, r=50),
    xaxis=dict(
        title='Month',
        tickformat='%b %Y',  # Showing month-year format
        showgrid=True,
        gridcolor='lightgray',
        tickmode='array',  
        tickvals=growth_data['month'].unique()  # only unique months
    ),
    yaxis=dict(
        title='Growth Rate (%)',
        showgrid=True,
        gridcolor='lightgray'
    ),
    showlegend=True  
)

# Find the peak growth points (max growth rate per product category)
peak_growth = growth_data.loc[growth_data.groupby('product_category')['sales_growth_rate_display'].idxmax()]

# Manually assigning colors from the color palette
color_palette = px.colors.qualitative.Set1
color_map = {category: color_palette[i % len(color_palette)] for i, category in enumerate(peak_growth['product_category'].unique())}

# Added smaller, bordered circles that match the line color
for category in peak_growth['product_category'].unique():
    category_data = peak_growth[peak_growth['product_category'] == category]
    category_color = color_map[category]  

    fig.add_trace(go.Scatter(
        x=category_data['month'],
        y=category_data['sales_growth_rate_display'],
        mode='markers',
        marker=dict(
            color=category_color,  # Match the marker color to the line color
            size=6,  
            symbol='circle',
            line=dict(width=1, color='black')  
        ),
        name=f'Peak Growth ({category})',
        showlegend=False  
    ))


fig.update_traces(
    line=dict(width=2)  
)

# Set hover template to include product category and correctly formatted percentage
for trace in fig.data:
    trace.update(
        hovertemplate=(
            '<b>Product Category: %{fullData.name}</b><br>'  # Product category from the trace name
            'Month: %{x|%b %Y}<br>'  # Month in 'Month Year' format
            'Growth Rate: %{y:.2f}%'  # Growth rate shown as a percentage with two decimals
        )
    )

# Show the plot
fig.show()


In [520]:
fig.write_html("Monthly_product_analysis.html")

---

### Interpretation
The data shows month-over-month growth rates for various product categories, with some categories experiencing notable fluctuations. **Branded** products show a peak growth rate in March, while **Packaged Chocolate** reaches its highest growth in April, followed by a decline in the following months. This indicates a possible seasonal or event-driven demand pattern for these categories, possibly tied to holidays or specific promotions. Other categories, such as **Loose Tea** and **Flavours** etc, maintain more stable growth rates, suggesting steady demand throughout the period without significant spikes or drops.

---

### Business Insights
For shop owners, these insights can guide inventory and marketing strategies to align with demand cycles. Branded products and Packaged Chocolate, with their strong peaks in March and April, respectively, may benefit from targeted promotional campaigns or increased stock levels during these months to fully capitalize on heightened demand. Conversely, for consistently growing categories like Loose Tea and Flavours, shop owners could consider implementing loyalty programs or subscription-based options to encourage repeat purchases and build a loyal customer base. By adjusting stock and promotional efforts according to these growth patterns, shop owners can optimize sales and improve inventory management throughout the year.

---

# **Pricing and Revenue Insights**

### 9. What is the average transaction value per store and how does it compare across different stores?

In [521]:
# Calculate the average transaction value per store location
average_transaction_value = df.groupby('store_location')['total_revenue'].mean().reset_index()

In [522]:
average_transaction_value

store_location  total_revenue
0          Astoria       4.589891
1   Hell's Kitchen       4.661696
2  Lower Manhattan       4.814726

In [523]:
# Creating the bar chart for total revenue per store location
fig = px.bar(
    average_transaction_value, 
    x='store_location', 
    y='total_revenue', 
    title='<b>Average Transaction Value per Store</b>',
    labels={'total_revenue': 'Average Transaction Value', 'store_location': 'Store Location'},
    color='store_location', 
    color_discrete_sequence=px.colors.qualitative.Set1,  # Set consistent color palette
    text='total_revenue'  # Display total revenue above the bars
)

fig.update_layout(
    title_font=dict(size=22, family='Arial', color='black'),
    font=dict(size=14, family='Arial'),
    height=500,
    width=900,
    plot_bgcolor='white',  
    margin=dict(t=60, b=50, l=50, r=50),
    showlegend=False,  
    xaxis=dict(
        title='Store Location',
        showgrid=True,
        gridcolor='lightgray',  
    ),
    yaxis=dict(
        title='Average Transaction Value',
        showgrid=True,
        gridcolor='lightgray',  
    )
)

# Enhance text visibility and bar appearance
fig.update_traces(
    texttemplate='%{text:.3s}', 
    textposition='outside',  
    marker_line_width=1.5,  
    textfont=dict(size=12, color='black', family='Arial', weight='bold')  
)

# Show the plot
fig.show()


In [524]:
fig.write_html("Avrg_trans_value.html")

---

**Interpretation of the Data**  
The bar chart compares the average transaction values across all three store locations: Astoria, Hell's Kitchen, and Lower Manhattan. Lower Manhattan  leads with the highest average transaction value at $4.81, followed by Hell's Kitchen at $4.66, and Astoria at $4.59. This indicates that customers in Hell's Kitchen tend to spend more per transaction than those in the other locations.

---

**Business Insights and Suggestions**  
Capitalize on Lower Manhattan, Introduce premium products or upselling strategies to further boost transaction value. Enhance Hell's Kitchen & Astoria, Consider exclusive offerings or promotions to encourage higher spending at these locations.

---

### 10.	How do sales of premium items (e.g., higher-priced coffees) compare to budget options across different store locations?

In [525]:
# Defining premium and budget items based on price thresholds
premium_threshold = df['unit_price'].mean() + df['unit_price'].std()

# Classify items as premium or budget
df['item_type'] = df['unit_price'].apply(lambda x: 'Premium' if x > premium_threshold else 'Budget')

# Group by store location and item type to calculate total sales
sales_comparison = df.groupby(['store_location', 'item_type'])['total_revenue'].sum().reset_index()

sales_comparison

store_location item_type  total_revenue
0          Astoria    Budget      212284.00
1          Astoria   Premium       19959.91
2   Hell's Kitchen    Budget      209773.95
3   Hell's Kitchen   Premium       26737.22
4  Lower Manhattan    Budget      207440.89
5  Lower Manhattan   Premium       22616.36

In [526]:
# Creating a side-by-side bar plot to visualize the comparison
fig = px.bar(
    sales_comparison, 
    x='store_location', 
    y='total_revenue', 
    color='item_type', 
    barmode='group',  # Ensures the bars are side by side
    title='<b>Sales Comparison of Premium vs Budget Items Across Store Locations</b>', 
    labels={'total_revenue': 'Total Sales', 'store_location': 'Store Location'},
    color_discrete_sequence=px.colors.qualitative.Set1
)

fig.update_layout(
    title_font=dict(size=20, family='Arial', color='black'),
    font=dict(size=12, family='Arial'),
    height=600,
    width=750,
    plot_bgcolor='white',
    margin=dict(t=60, b=50, l=50, r=50),
    xaxis=dict(
        title='Store Location',
        showgrid=True,
        gridcolor='lightgray',
    ),
    yaxis=dict(
        title='Total Sales',
        showgrid=True,
        gridcolor='lightgray',
    ),
    legend_title=dict(text='Item Category')  
)

# Enhance text visibility and bar appearance
fig.update_traces(
    texttemplate='%{y:.4s}', 
    textposition='outside',
    marker_line_width=1.5,
    textfont=dict(size=11, color='black', family='Arial', weight='bold')
)

# Show the plot
fig.show()


In [527]:
fig.write_html("PremiumVSBudget.html")

---

### **Interpretation of the Data**  
The chart reveals that budget items consistently outperform premium items in total sales across all locations. In Astoria, budget items generate $212.3k in sales, while premium items generate $19.96k. Hell's Kitchen and Lower Manhattan show similar trends, with budget items significantly outpacing premium sales.

---

### **Business Insights & Suggestions**  
Since budget items drive the majority of sales, focusing on maintaining their stock and promoting them further will maximize revenue. To boost premium item sales, targeted marketing or bundling them with budget items could help increase their appeal and attract more customers to higher-priced options.

---


# **Operational Efficiency**

### 11. What is the relationship between transaction time (hour of the day) and transaction size?

In [528]:
df.head(2)

transaction_id transaction_date transaction_time  transaction_qty  \
0               1       2023-01-01         07:06:11                2   
1               2       2023-01-01         07:08:56                2   

   store_id   store_location  product_id  unit_price product_category  \
0         5  Lower Manhattan          32         3.0           Coffee   
1         5  Lower Manhattan          57         3.1              Tea   

            product_type            product_detail transaction_datetime  \
0  Gourmet brewed coffee               Ethiopia Rg  2023-01-01 07:06:11   
1        Brewed Chai tea  Spicy Eye Opener Chai Lg  2023-01-01 07:08:56   

   transaction_hour  total_revenue    month day_of_week item_type  
0                 7            6.0  2023-01      Sunday    Budget  
1                 7            6.2  2023-01      Sunday    Budget

In [529]:
hourly_sales=df.groupby('transaction_hour')['transaction_qty'].sum().reset_index()

In [530]:
# Create a line plot for hourly sales quantity with a customized color scheme
fig = px.line(
    hourly_sales, 
    x='transaction_hour', 
    y='transaction_qty', 
    markers=True, 
    title='<b>Hourly Sales Quantity</b>', 
    labels={'transaction_hour': 'Hour of the Day', 'transaction_qty': 'Total Transaction Quantity'}
)

# Update layout 
fig.update_layout(
    title_font=dict(size=22, family='Arial', color='black'),
    font=dict(size=14, family='Arial'),
    height=600,
    width=1000,
    plot_bgcolor='white',
    margin=dict(t=80, b=60, l=50, r=50),
    legend=dict(
        title="Sales Data",
        x=0.85,  # Position the legend
        y=0.95,
        bgcolor='rgba(255, 255, 255, 0.7)',  
        bordercolor='black',
        borderwidth=1,
        font=dict(size=12, family='Arial') 
    ),
    xaxis=dict(
        title='Hour of the Day',
        tickmode='array',
        tickvals=list(range(24)),
        showgrid=True,
        gridcolor='lightgray',
        showline=True,
        linewidth=2,
        linecolor='black',
        ticks='outside',
    ),
    yaxis=dict(
        title='Total Transaction Quantity',
        showgrid=True,
        gridcolor='lightgray',
        showline=True,
        linewidth=2,
        linecolor='black',
        ticks='outside',
    )
)

# Enhance line and marker appearance with color, size, and style
fig.update_traces(
    line=dict(width=4, color='royalblue', dash='solid'),  # Solid line with a blue color
    marker=dict(size=8, color='red', symbol='circle', line=dict(width=2, color='black')),  # Red markers with black border
    name='Sales Quantity'  # Name for the legend
)
# Show the plot
fig.show()


In [531]:
fig.write_html("HourOfDayVSTotalTrans.html")

---

### **Interpretation of the Data**  
The line plot shows that transaction volumes peak from 8 AM to 10 AM, followed by a decline that stabilizes through the afternoon before tapering off in the evening. This indicates early mornings are the busiest times, with steady activity in the afternoon and a slowdown in the evening.

---

### **Business Insights & Suggestions**  
Shop owners should focus on increasing staffing and inventory during peak morning hours to manage high transaction volumes. Maintaining steady resources in the afternoon will ensure smooth operations. Evening promotions or offers could help boost sales and attract more customers during slower hours. Understanding these patterns will improve resource allocation and optimize operations.

---

### 12.	Which products have the highest repeat purchase rate, indicating customer loyalty and demand consistency?

In [532]:
df.head(2)

transaction_id transaction_date transaction_time  transaction_qty  \
0               1       2023-01-01         07:06:11                2   
1               2       2023-01-01         07:08:56                2   

   store_id   store_location  product_id  unit_price product_category  \
0         5  Lower Manhattan          32         3.0           Coffee   
1         5  Lower Manhattan          57         3.1              Tea   

            product_type            product_detail transaction_datetime  \
0  Gourmet brewed coffee               Ethiopia Rg  2023-01-01 07:06:11   
1        Brewed Chai tea  Spicy Eye Opener Chai Lg  2023-01-01 07:08:56   

   transaction_hour  total_revenue    month day_of_week item_type  
0                 7            6.0  2023-01      Sunday    Budget  
1                 7            6.2  2023-01      Sunday    Budget

In [533]:
# Calculating repeat purchase rate for each product type
repeat_purchase_rate_type = df.groupby('product_type').agg(
    total_purchases=('transaction_qty', 'sum'),
    unique_customers=('transaction_id', 'nunique')
).reset_index()

# Calculating repeat purchases (total purchases - unique customers)
repeat_purchase_rate_type['repeat_purchases'] = repeat_purchase_rate_type['total_purchases'] - repeat_purchase_rate_type['unique_customers']

# Calculating the repeat purchase rate
repeat_purchase_rate_type['repeat_purchase_rate'] = repeat_purchase_rate_type['repeat_purchases'] / repeat_purchase_rate_type['total_purchases']

# Sorting by repeat purchase rate
repeat_purchase_rate_type = repeat_purchase_rate_type.sort_values(by='repeat_purchase_rate', ascending=False)

# Selecting the top 10 product types by repeat purchase rate
top_repeat_type = repeat_purchase_rate_type.head(10)


top_repeat_type

product_type  total_purchases  unique_customers  \
28       Sugar free syrup             2905              1811   
6       Brewed herbal tea            17328             11245   
3        Brewed Black tea            17462             11350   
13  Gourmet brewed coffee            25973             16912   
5        Brewed Green tea             8697              5671   
22  Organic brewed coffee            13012              8489   
25  Premium brewed coffee            12431              8135   
4         Brewed Chai tea            26250             17183   
26          Regular syrup             7606              4979   
17          Hot chocolate            17457             11468   

    repeat_purchases  repeat_purchase_rate  
28              1094              0.376592  
6               6083              0.351050  
3               6112              0.350017  
13              9061              0.348862  
5               3026              0.347936  
22              4523              0.347602  
25              4296              0.345588  
4               9067              0.345410  
26              2627              0.345385  
17              5989              0.343072

In [534]:
# Calculating repeat purchase rate for each product category
repeat_purchase_rate_category = df.groupby('product_category').agg(
    total_purchases=('transaction_qty', 'sum'),
    unique_customers=('transaction_id', 'nunique')
).reset_index()

# Calculating repeat purchases (total purchases - unique customers)
repeat_purchase_rate_category['repeat_purchases'] = repeat_purchase_rate_category['total_purchases'] - repeat_purchase_rate_category['unique_customers']

# Calculating the repeat purchase rate
repeat_purchase_rate_category['repeat_purchase_rate'] = repeat_purchase_rate_category['repeat_purchases'] / repeat_purchase_rate_category['total_purchases']

# Sorting by repeat purchase rate
repeat_purchase_rate_category = repeat_purchase_rate_category.sort_values(by='repeat_purchase_rate', ascending=False)

# Selecting the top 10 product categories by repeat purchase rate
top_repeat_category = repeat_purchase_rate_category.head(10)

# Displaying the top product categories by repeat purchase rate
top_repeat_category

product_category  total_purchases  unique_customers  repeat_purchases  \
5            Flavours            10511              6790              3721   
8                 Tea            69737             45449             24288   
2              Coffee            89250             58416             30834   
4  Drinking Chocolate            17457             11468              5989   
3        Coffee beans             1828              1753                75   
1             Branded              776               747                29   
0              Bakery            23214             22796               418   
6           Loose Tea             1210              1210                 0   
7  Packaged Chocolate              487               487                 0   

   repeat_purchase_rate  
5              0.354010  
8              0.348280  
2              0.345479  
4              0.343072  
3              0.041028  
1              0.037371  
0              0.018006  
6              0.000000  
7              0.000000

In [535]:
# Convert repeat purchase rates to percentages out of 100
top_repeat_type['repeat_purchase_rate'] = [rate * 100 for rate in top_repeat_type['repeat_purchase_rate']]
top_repeat_category['repeat_purchase_rate'] = [rate * 100 for rate in top_repeat_category['repeat_purchase_rate']]

# Generate colors from the Set1 color scale
set1_colors = px.colors.qualitative.Set1

# Create a subplot with 1 row and 2 columns for the bar charts
fig = make_subplots(
    rows=1, cols=2,
    subplot_titles=(
        "<b>Repeat Purchase Rate by Product Type</b>",
        "<b>Repeat Purchase Rate by Product Category</b>"
    ),
    horizontal_spacing=0.15
)

# Bar chart for Product Type
fig.add_trace(
    go.Bar(
        x=top_repeat_type['product_type'],
        y=top_repeat_type['repeat_purchase_rate'],
        text=[f"{rate:.2f}%" for rate in top_repeat_type['repeat_purchase_rate']],
        textposition='auto',
        textfont=dict(
            size=14,
            family='Arial',
            weight='bold'  
        ),
        marker=dict(
            color=[set1_colors[i % len(set1_colors)] for i in range(len(top_repeat_type))],  # Set color from Set1
            line=dict(color='rgba(0,0,0,0)', width=0)  
        ),
        hovertemplate='<b>%{x}</b><br>Rate: %{y:.2f}%<extra></extra>',
        name="Product Type"
    ),
    row=1, col=1
)

# Bar chart for Product Category
fig.add_trace(
    go.Bar(
        x=top_repeat_category['product_category'],
        y=top_repeat_category['repeat_purchase_rate'],
        text=[f"{rate:.2f}%" for rate in top_repeat_category['repeat_purchase_rate']],
        textposition='auto',
        textfont=dict(
            size=14,
            family='Arial',
            weight='bold'  
        ),
        marker=dict(
            color=[set1_colors[i % len(set1_colors)] for i in range(len(top_repeat_category))],  
            line=dict(color='rgba(0,0,0,0)', width=0) 
        ),
        hovertemplate='<b>%{x}</b><br>Rate: %{y:.2f}%<extra></extra>',
        name="Product Category"
    ),
    row=1, col=2
)

fig.update_layout(
    title_text="<b>Top 10 Product Types and Categories by Repeat Purchase Rate</b>",
    title_font=dict(size=20, family='Arial', color='black'),
    font=dict(size=12, family='Arial'),
    height=450,
    width=1500,
    plot_bgcolor='white',  
    margin=dict(t=80, b=80, l=40, r=40),
    xaxis=dict(
        showgrid=True, 
        gridcolor='lightgrey', 
        zeroline=False,
        tickangle=-45,  # Rotate x-axis labels for better readability
        tickfont=dict(size=12, family='Arial', color='black'),  # Increase font size for better visibility
        ticks='outside',  # Place ticks outside for better emphasis
        tickwidth=2  
    ),
    yaxis=dict(
        showgrid=True, 
        gridcolor='lightgrey'
    ),
    xaxis2=dict(
        showgrid=True, 
        gridcolor='lightgrey', 
        zeroline=False,
        tickangle=-45,  
        tickfont=dict(size=12, family='Arial', color='black'),  
        ticks='outside',  
        tickwidth=2 
    ),
    yaxis2=dict(
        showgrid=True, 
        gridcolor='lightgrey'
    ),
    showlegend=False  
)

# Show the plot
fig.show()


---

**Interpretation of the Data**  
The bar charts titled "Repeat Purchase Rate by Product Type" and "Repeat Purchase Rate by Product Category" reveal key insights into customer loyalty. Sugar Free Syrup, Brewed Herbal Tea, and Brewed Black Tea exhibit the highest repeat purchase rates, with values around 35%. Product categories like Flavours, Tea, and Coffee also show strong repeat purchase rates, suggesting a high level of customer retention for these items.

---

**Business Insights and Suggestions**  
Focusing on high-repeat purchase products like Sugar Free Syrup, Brewed Herbal Tea, and Flavours can help drive customer loyalty. Ensuring consistent stock and promotion of these products will likely increase repeat purchases, enhancing overall sales. Introducing loyalty programs or discounts on these products could further improve customer retention, while offering new variations or seasonal releases can attract new customers and keep the existing ones engaged.

---